orchagent/vnetorch.cpp

```diff
@@ -2004,6 +2014,160 @@ bool VNetRouteOrch::delOperation(const Request& request)
+        assert(false);
+        return;
+    }
+    for (const auto& route : observerEntry->second.routeTable)
+    {
+        VNetNextHopUpdate update = {
+            DEL_COMMAND,
+            std::get<0>(route.second), // vnet name
+            dstAddr, // destination
+            route.first, // prefix
+            std::get<1>(route.second) // nexthop
+        };
+        observer->update(SUBJECT_TYPE_NEXTHOP_CHANGE, reinterpret_cast<void*>(&update));
+    }
+    next_hop_observers_.erase(observerEntry);
+}
+
+void VNetRouteOrch::addRoute(const std::string& vnet, const IpPrefix& ipPrefix, const nextHop& nh)
+{
+    SWSS_LOG_ENTER();
+
+    for (auto & next_hop_observer : next_hop_observers_)
+    {
+        if (ipPrefix.isAddressInSubnet(next_hop_observer.first))
+        {
+            next_hop_observer.second.routeTable.emplace(ipPrefix, std::make_tuple(vnet, nh));
+            for (auto& observer : next_hop_observer.second.observers)
+            {
+                VNetNextHopUpdate update = { 
+                    SET_COMMAND, 
+                    vnet,
+                    next_hop_observer.first, 
+                    ipPrefix, 
+                    nh
+                };
+                observer->update(SUBJECT_TYPE_NEXTHOP_CHANGE, reinterpret_cast<void*>(&update));
+            }
+        }
+    }
+    syncd_routes_.emplace(ipPrefix, std::make_tuple(vnet, nh));
+}
+
+void VNetRouteOrch::delRoute(const IpPrefix& ipPrefix)
+{
+    SWSS_LOG_ENTER();
+
+    auto route_itr = syncd_routes_.find(ipPrefix);
+    if (route_itr == syncd_routes_.end())
+    {
+        SWSS_LOG_ERROR("Failed to find route %s.", ipPrefix.to_string().c_str());
+        assert(false);
+        return;
+    }
+    auto next_hop_observer = next_hop_observers_.begin();
+    while(next_hop_observer != next_hop_observers_.end())
+    {
+        if (ipPrefix.isAddressInSubnet(next_hop_observer->first))
+        {
+            auto itr = next_hop_observer->second.routeTable.find(ipPrefix);
+            if ( itr == next_hop_observer->second.routeTable.end())
+            {
+                SWSS_LOG_ERROR(
+                    "Failed to find any ip(%s) belong to this route(%s).", 
+                    next_hop_observer->first.to_string().c_str(),
+                    ipPrefix.to_string().c_str());
+                assert(false);
+                continue;
+            }
+            for (auto& observer : next_hop_observer->second.observers)
+            {
+                VNetNextHopUpdate update = {
+                    DEL_COMMAND,
+                    std::get<0>(itr->second), // vnet name
+                    next_hop_observer->first, // destination
+                    itr->first, // prefix
+                    std::get<1>(itr->second) // nexthop
+                };
+                observer->update(SUBJECT_TYPE_NEXTHOP_CHANGE, reinterpret_cast<void*>(&update));
```

Comment:
do you need to check if the removed route is a 'better' match for the current observers?

Reply:
Let's discuss it at teams.

Reply:
Done~ Please review the newest.